feat(web-analytics): adding marketing settings page to setup bq

jabahamondes opened the pull request:
## Problem

Currently it's very hardcoded how we connect the tables. So now there a UI in marketing settings that allows you to connect and map any table.

This also include refactoring the currency into a global setting now and making more reusable the links logic for the source creation url.

## Changes

- Introduced new marketing tab in web analytics with relevant routing and insights.
- Updated constants and scene configurations to include marketing features.
- Enhanced WebAnalyticsDashboard to conditionally display marketing costs based on feature flags.
- Adjusted logic to handle marketing product tab and its associated breadcrumbs.
- Updated WebPropertyFilters to hide filters when on the marketing tab


## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've added or updated the docs
- [ ] I've reached out for help from the docs team
- [x] No docs needed for this change

## How did you test this code?

Manually
<img width="1202" alt="image" src="https://github.com/user-attachments/assets/203ab086-e8a9-4ccc-990a-3648b4fa9b9a" />

<img width="1241" alt="image" src="https://github.com/user-attachments/assets/bc9c89e5-4740-435e-9235-442f3398cf69" />

We only show the sources that have the required fields mapped
<img width="1197" alt="image" src="https://github.com/user-attachments/assets/cec23355-2714-4d6b-bd43-b009b23d62fc" />
<img width="1202" alt="image" src="https://github.com/user-attachments/assets/751ba5dd-1311-405f-83e6-34c6d65c1623" />

Commit:
feat(web-analytics): add marketing insights and routing

- Introduced new marketing tab in web analytics with relevant routing and insights.
- Updated constants and scene configurations to include marketing features.
- Enhanced WebAnalyticsDashboard to conditionally display marketing costs based on feature flags.
- Adjusted logic to handle marketing product tab and its associated breadcrumbs.
- Updated WebPropertyFilters to hide filters when on the marketing tab.

Commit:
Merge branch 'master' into feat/creating-marketing-template

Commit:
refctor: Reverting temp code

Commit:
Merge branch 'master' into feat/creating-marketing-template

Commit:
feat(web-analytics): add marketing tab to dashboard with beta label
- Introduced a new marketing tab in the Web Analytics Dashboard, displayed conditionally based on feature flags.
- The tab includes a "Beta" label for clarity.

Commit:
Adding comment

Commit:
fix(web-analytics): update marketing campaign breakdown tile priority and remove date filters
- Changed priority of the MARKETING_CAMPAIGN_BREAKDOWN tile to 1.
- Removed unnecessary date filters from the web analytics logic.

Commit:
feat(web-analytics): implement marketing analytics settings and configuration
- Added MarketingAnalyticsConfig and MarketingAnalyticsSchema to manage marketing analytics settings.
- Introduced BaseCurrency and CurrencyDropdown components for currency selection.
- Created NonNativeExternalDataSourceConfiguration for managing external data sources.
- Updated routing and scene logic to include marketing analytics tab in Data Management.
- Enhanced schema definitions in schema.json for marketing analytics.

Commit:
fix(schema): remove pageviews from marketing analytics schema
- Deleted the `pageviews` field from both schema.json and schema-general.ts as it is no longer needed in the marketing analytics context.

Commit:
Merge branch 'master' into feat/adding-marketing-settings

Commit:
feat(web-analytics): Conditional rendering of WebPropertyFilters based on productTab

- Added productTab to the state in WebAnalyticsFilters and WebAnalyticsModal.
- Updated WebPropertyFilters to render conditionally based on productTab, excluding it for the MARKETING tab.
- Adjusted campaign costs breakdown query to use the correct property name for campaign.

Commit:
Merge branch 'feat/creating-marketing-template' into feat/adding-marketing-settings

Commit:
feat(marketing-analytics): Introduce Marketing Analytics configuration and schema

- Added  to  for defining marketing analytics data structure.
- Updated  to export  type.
- Integrated marketing analytics settings in  and .
- Enhanced  to manage mappings for marketing analytics.
- Implemented  model and serializer for handling marketing analytics configurations.
- Added migration for  to support new settings.
- Updated logic in  to accommodate dynamic marketing data queries.
- Adjusted UI components to reflect changes in marketing analytics settings and configurations.

Commit:
Merge branch 'master' into feat/adding-marketing-settings

github-actions[bot] committed:
Update query snapshots

Commit:
fix: ensure field mappings are removed only if they exist in NonNativeExternalDataSourceConfiguration

- Updated removeTableMapping function to check for existing source mappings before attempting to remove field mappings, preventing potential errors when a tableId has no mappings.

Commit:
Merge branch 'feat/adding-marketing-settings' of github.com:PostHog/posthog into feat/adding-marketing-settings

Commit:
fix: changing to area chart

Commit:
Merge branch 'master' into feat/adding-marketing-settings

Commit:
fix: correct serializer variable name for marketing analytics config in TeamSerializer

Updated the variable name from `serializer` to `marketing_serializer` for clarity and accuracy when validating and saving the marketing analytics configuration. This change ensures that the correct serializer instance is used, preventing potential validation errors.

Commit:
Fixing CI

Commit:
Merge branch 'master' into feat/adding-marketing-settings

Commit:
fix: update snapshot for verifiedDomainsLogic test to include marketing_analytics_config

Commit:
fix: remove conflicting migration file before recreating

Commit:
Merge master and resolve migration conflicts

Commit:
feat(migrations): add TeamMarketingAnalyticsConfig model with base_currency and sources_map fields

Commit:
feat: removing duplicated file

Commit:
feat(products): add getTreeItemsMetadata function for actions and revenue settings

Commit:
Merge master and resolve conflicts

Commit:
Update migration file number and dependencies

jabahamondes commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -746,6 +756,165 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
     }),
     selectors(({ actions, values }) => ({
+        // Helper functions for dynamic marketing analytics
+        createMarketingDataWarehouseNodes: [
```
Most of this logic will be moved to a query runner file

rafaeelaudibert replied:
Yes, that's the approach I took with Revenue Analytics eventually and it's much much nicer, good call

jabahamondes commented:
There are improvements that will come in the future. This is the PR that will continue with some of them: https://github.com/PostHog/posthog/pull/new/feat/adding-s3-and-currency-support (Draft)

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert commented on frontend/src/queries/schema/schema-general.ts:
```diff
@@ -3107,3 +3107,29 @@ export interface EventsHeatMapStructuredResult {
+    required: boolean
+}
+
+export const MARKETING_ANALYTICS_SCHEMA: Record<string, MarketingAnalyticsSchemaField> = {
+    campaign_name: { type: ['string'], required: true },
+    total_cost: { type: ['float', 'integer'], required: true },
+    clicks: { type: ['integer', 'number', 'float'], required: true },
+    impressions: { type: ['integer', 'number', 'float'], required: true },
+    date: { type: ['datetime', 'date'], required: true },
+    source_name: { type: ['string'], required: false },
+}
```
Why is this typed like this rather than a more normal TS-based validation? We're using `ajv` in some places, you could use that instead

jabahamondes replied:
Yes, it's so it can be mapped by the user and I can added the columns I want. But I will move it from there as you suggested 🫡

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/BaseCurrency.tsx:
```diff
@@ -0,0 +1,27 @@
+import { marketingAnalyticsSettingsLogic } from '../../logic/marketingAnalyticsSettingsLogic'
+import { CurrencyDropdown } from './CurrencyDropdown'
+
+export function BaseCurrency(): JSX.Element {
```
In case you agree with my general comment, you'd then reuse the `BaseCurrency` component created for revenue analytics - of course, we'd need to update *where* we're saving the data so that it's in the team, but that's trivial.

jabahamondes replied:
Agree! Should be fixed with the refactor

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/CurrencyDropdown.tsx:
```diff
@@ -0,0 +1,67 @@
+const OPTIONS_FOR_IMPORTANT_CURRENCIES: LemonSelectOption<CurrencyCode>[] = IMPORTANT_CURRENCIES.map(optionFromCurrency)
+const OPTIONS_FOR_OTHER_CURRENCIES: LemonSelectOption<CurrencyCode>[] = OTHER_CURRENCIES.map(optionFromCurrency)
+
+export const CurrencyDropdown = ({ value, onChange, visible, size }: CurrencyDropdownProps): JSX.Element => {
```
A lot of duplication here, even more reason to just store this all in the same place and reuse the components

jabahamondes replied:
Agree! Should be fixed with the refactor

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/ExternalDataSourceConfiguration.tsx:
```diff
@@ -0,0 +1,97 @@
+
+const VALID_REVENUE_SOURCES: ExternalDataSource['source_type'][] = ['BigQuery']
+
+export function ExternalDataSourceConfiguration({
```
I wonder if we should just have a generic `ExternalDataSourceConfiguration` that receives the `VALID_SOURCES` param instead. You would configure *where* to save the values based on the logic props - which you'd then mount before initializing this component.

rafaeelaudibert replied:
Oh, maybe this whole component is unnecessary since you aren't using and instead using the `NonNative` one? It's not clear to me

jabahamondes replied:
This i might refactor in another PR where I add self managed resources if you are ok with that 👌

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/ExternalDataSourceConfiguration.tsx:
```diff
@@ -0,0 +1,97 @@
+
+import { marketingAnalyticsSettingsLogic } from '../../logic/marketingAnalyticsSettingsLogic'
+
+const VALID_REVENUE_SOURCES: ExternalDataSource['source_type'][] = ['BigQuery']
```
Make this more reasonable for this component

```suggestion
const VALID_MARKETING_SOURCES: ExternalDataSource['source_type'][] = ['BigQuery']
```

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/ExternalDataSourceConfiguration.tsx:
```diff
@@ -0,0 +1,97 @@
+                                ref={buttonRef}
+                                type="primary"
+                                onClick={() => {
+                                    router.actions.push(urls.pipelineNodeNew(PipelineStage.Source, { kind: 'stripe' }))
```
You dont want Stripe by default, update to BigQuery

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/ExternalDataSourceConfiguration.tsx:
```diff
@@ -0,0 +1,97 @@
+                            )
+                        },
+                    },
+                    {
+                        key: 'revenue_analytics_enabled',
+                        title: 'Enabled?',
+                        render: (_, item: ExternalDataSource) => {
+                            return (
+                                <LemonSwitch
+                                    checked={item.revenue_analytics_enabled}
+                                    onChange={(checked) =>
+                                        updateSource({ ...item, revenue_analytics_enabled: checked })
+                                    }
+                                />
+                            )
+                        },
+                    },
```
I assume you want a new column rather than `revenue_analytics_enabled`, right?

jabahamondes replied:
I won't need this actually

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/settings/NonNativeExternalDataSourceConfiguration.tsx:
```diff
@@ -0,0 +1,257 @@
+                            )
+                        },
+                    },
+                    ...Object.keys(MARKETING_ANALYTICS_SCHEMA).map((column) => ({
+                        key: column,
+                        title: `${column.replace(/_/g, ' ').replace(/\b\w/g, (l) => l.toUpperCase())}${
+                            MARKETING_ANALYTICS_SCHEMA[column].required ? ' (*)' : ''
+                        }`,
+                        render: (_: any, item: any) => renderColumnMappingDropdown(item, column),
+                    })),
```
That's interesting, I now see why you built it that way, but I dont think that should live on `schema` though, maybe we can put it somewhere else?

rafaeelaudibert commented on posthog/models/team/team.py:
```diff
@@ -408,6 +408,13 @@ def revenue_analytics_config(self):
         config, _ = TeamRevenueAnalyticsConfig.objects.get_or_create(team=self)
         return config
 
+    @cached_property
+    def marketing_analytics_config(self):
+        from .team_marketing_analytics_config import TeamMarketingAnalyticsConfig
+
+        config, _ = TeamMarketingAnalyticsConfig.objects.get_or_create(team=self)
+        return config
```
BTW, you might wanna backfill these eventually for older teams, this way you can almost-always assume they exist. Django is kinda bad for this kind of behavior, but this has been enough for me.

rafaeelaudibert reviewed (commented):
Above all, are you positive we need a separate currency config? I believe we could get away with just one global currency that we use to display data everywhere we have currency.

That would imply moving from `RevenueAnalyticsConfig` to `Team` but that's simple to do, how do you feel about that?

Of course, we can just ship your code as is and we can change this in the future, but I'd like you to know I feel strong about merging them eventually, and we'll need to update both codebases to achieve that

Commit:
refactor: currency refactor, moving into the entire team. Creating a migration to use revenue analytics default value.

Commit:
refactor: renaming

Commit:
refactor: stripe kind leak

Commit:
refactor: remove ExternalDataSourceConfiguration component and related actions

Deleted the ExternalDataSourceConfiguration component and removed its associated updateSource action from marketingAnalyticsSettingsLogic. This streamlines the codebase by eliminating unused functionality.

Commit:
refactor: removing unused field

Commit:
refactor: move MARKETING_ANALYTICS_SCHEMA to utils and clean up imports

- Moved MARKETING_ANALYTICS_SCHEMA definition to a new utils file for better organization.
- Updated imports in relevant files to reflect the new location.
- Removed the old schema definition from schema-general.ts to streamline the codebase.

Commit:
refactor: update pipelineNodeNew to use source instead of kind

- Changed the  function to accept  instead of  for better clarity and consistency across the codebase.
- Updated all relevant calls to  to reflect this change, ensuring that the source is now passed correctly.
- Removed unnecessary button reference in  for cleaner code.

Commit:
Merge branch 'master' into feat/adding-marketing-settings

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

Commit:
Merge branch 'master' into feat/adding-marketing-settings

Commit:
fix: fixing tests

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-marketing-settings

Commit:
Merge branch 'master' into feat/adding-marketing-settings

Commit:
fix(test_utils): add base_currency to team insert query in PostgreSQL

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert commented on frontend/__snapshots__/scenes-other-settings--settings-project--light.png:
Follow-up PR: base currency should probably rank slightly higher on this list

Commit:
fix: merging migrations

Commit:
Merge branch 'feat/adding-marketing-settings' of github.com:PostHog/posthog into feat/adding-marketing-settings

rafaeelaudibert commented on posthog/models/team/team_revenue_analytics_config.py:
```diff
@@ -1,27 +1,18 @@
 # and therefore using the exact same primary key as the Team model.
 class TeamRevenueAnalyticsConfig(models.Model):
     team = models.OneToOneField(Team, on_delete=models.CASCADE, primary_key=True)
-    base_currency = models.CharField(max_length=3, choices=CURRENCY_CODE_CHOICES, default=CurrencyCode.USD.value)
```
As mentioned above, dont remove this for now, running workers will crash when this is removed, we need to: first stop using it, then update this

rafaeelaudibert reviewed (changes requested):
We need to change our migrations slightly, the way they're setup now it'll crash our servers for a while until we've deployed the new version which isnt ideal. I know this is a huge PR, but it's also a big change, and we need to be careful with the migrations.

jabahamondes commented on frontend/src/scenes/settings/SettingsMap.tsx:
```diff
@@ -242,11 +248,6 @@ export const SETTINGS_MAP: SettingSection[] = [
         id: 'environment-revenue-analytics',
         title: 'Revenue analytics',
         settings: [
-            {
-                id: 'revenue-base-currency',
-                title: 'Revenue base currency',
-                component: <RevenueBaseCurrencySettings />,
-            },
```
I'll keep it 👌

lricoy commented:
+1 to @rafaeelaudibert comments. I think this will go easier if you move the schema changes and migration to its own PR and get it merged first. 

Then we can cut some slack to the things that are behind the FF so you can get more feedback on them.

Commit:
feat(currency): add DEFAULT_CURRENCY constant and update imports across components; remove unused MarketingAnalyticsSchema from schema-general

jabahamondes commented on frontend/src/queries/nodes/WebOverview/WebOverview.tsx:
```diff
@@ -16,7 +16,7 @@ import { urls } from 'scenes/urls'
 import {
     AnyResponseType,
-    CurrencyCode,
+    DEFAULT_CURRENCY,
```
Moved!

rafaeelaudibert commented on posthog/migrations/0760_remove_teammarketinganalyticsconfig_base_currency_and_more.py:
```diff
@@ -0,0 +1,270 @@
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("posthog", "0752_organization_members_can_invite"),
+        ("posthog", "0759_add_errortrackingsymbolset_last_used_index"),
+    ]
```
Just 0759 is needed here

rafaeelaudibert reviewed (approved):
LGTM, let's see if we can merge this 🤞🏻

Commit:
feat: revert code

Commit:
fix: update dependency

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert commented on posthog/models/team/team.py:
```diff
@@ -401,13 +411,27 @@ def aggregate_users_by_distinct_id(self) -> bool:
     revenue_tracking_config = models.JSONField(null=True, blank=True)
 
+    # Consolidated base currency for all analytics (revenue, marketing, etc.)
+    base_currency = models.CharField(
+        max_length=3,
+        choices=CURRENCY_CODE_CHOICES,
+        default=DEFAULT_CURRENCY,
+    )
```
You'll need `null=True` here to satisfy CI
```suggestion
    base_currency = models.CharField(
        max_length=3,
        choices=CURRENCY_CODE_CHOICES,
        default=DEFAULT_CURRENCY,
        null=True,
        blank=True,
    )
```

rafaeelaudibert replied:
Then regenerate migration
